cypress/support/PersistentLogs.js

```diff
@@ -16,15 +21,98 @@ console.log = function (message) {
   logEntries += message + '\r\n'
 }
 
-// This will be called by the test list manager to register each test case
-// in the order the tests will be run in.
-Cypress.PersistentLogs = {}
-Cypress.PersistentLogs.RegisterTestCase = (testGroupName, testName) => {
-  testCaseRegistry.push(`${testGroupName}/${testName}`)
-}
+// Write ONLY to the true console.log and NOT our persistent log.
+// Uncomment to use this for debugging purposes.
+// function consoleLogOnly() { originalConsolLog.apply(console, arguments) }
+
+// -----------------------------------------------------------------------------
+// Functions that add additional log entries related to the stage of the test 
+// suite and to persist the captured logs at the appropriate time.
+// -----------------------------------------------------------------------------
+
+// FINAL LOGGING SUPPORT FOR PREVIOUS TEST CASE
+// In an ideal world this code would be in an afterEach call, however, since 
+// there will be other afterEach functions that will be called after ours and 
+// we want to ensure the logging we do here is the final logging for the test 
+// case.
+//
+// Also important to notice is that this results in a call to cy.writeFile
+// which will execute in a different time sequence than all of the other 
+// JavaScript in this code block. Thus in this code block we cannot run any
+// true `beforeEach` code that is related to the current test case that is
+// just about to start.
+// FINAL LOGGING SUPPORT FOR PREVIOUS TEST CASE
+beforeEach(function() {
+  LogTestState()
+  FlushLogEntries()
+})
+
+// TRUE `beforeEach` CODE FOR THE TEST CASE THAT IS JUST NOW STARTING
+// This MUST come after the prior `beforeEach` code block and cannot
+// be combined with that other code block due to the `cy.writeFile`
+// call that it contains.
+beforeEach(function() {
+  console.log('')
+  console.log(`******************** ${GetFullTestSuiteTitle(this.currentTest)} ************************************************************`)
+
+  test = this.currentTest
+  let title = GetSuiteRootTitle(this.currentTest)
+  if (title !== suiteTitle) {
+    suiteTitle = title
+    logFileName = `./results/cypress/${suiteTitle}.${Cypress.moment().format("YY.MM.DD.HH.mm.ss..SSS")}.log`
+  }
+})
 
 // After each test case ends...
-afterEach(() => {
-  let logFileName = `./results/cypress/${testCaseRegistry[indexRegistry++]}.${Cypress.moment().format("YY.MM.DD.HH.mm.ss..SSS")}.log`
-  cy.writeFile(logFileName, logEntries).then(() => logEntries = '')
+afterEach(() => { WriteLogEntries() })
+
+after(() => { 
+  LogTestState()
+  FlushLogEntries() 
 })
+
+// -----------------------------------------------------------------------------
+// Support functions.
+// -----------------------------------------------------------------------------
+
+function LogTestState()
+{
+  if (!test) { return false }
```

Comment:
Looks like it relies on a global variable. Can you try to achieve this without globals?

Seems like anyone could overwrite this to something arbitrary like `test = { foo: "bar" }` and this line would pass and try to use it.

Reply:
It is only global to the scope of this code file, it is not exported. It is defined as:
> var test
> 
at the beginning of the file.

Reply:
Yes, which is a global ... and also uses `var` both things we don't want.

Reply:
I wrapped it all in a nameless function scope and removed the var/globals.